### PR TITLE
bpo-30576 : Add HTTP compression support to http.server

### PR DESCRIPTION
**Pull request by PierreQuentel**
The Pull Request add support of HTTP compression to the built-in HTTP server. For some files (html, css, js, json...), compression reduces network load by sending the gzipped content instead of the original content. Most browsers support it.

The main change in is method `send_head()` of `SimpleHTTPRequestHandler`. If the request includes the header `Accept-Encoding` and its content includes "gzip", and if the content type determined by the file extension is in a list of compressed types, then the response includes the header `Content-Encoding` set to "gzip", the content type is set to that of the gzipped content, and the file object returned by `send_head()` points to the gzipped content.

If the original file is small, its content is loaded in memory and gzipped, the file object is and instance of `io.BytesIO`. If it is big, the gzipped content is stored in a temporary file.

The PR changes module __mimetypes__ to add the mapping of extension ".json" to content type "application/json".

https://bugs.python.org/issue30576

### CODE REVIEW

File: Doc/whatsnew/3.7.rst
Comment (matrixise) [inline Doc/whatsnew/3.7.rst]: could be in an other PR.

File: Lib/http/server.py
Comment (matrixise) [inline Lib/http/server.py]: remove this line

File: Lib/mimetypes.py
Comment (matrixise) [inline Lib/mimetypes.py]: in an other PR ? it's out of context

File: Lib/test/test_httpservers.py
Comment (matrixise) [inline Lib/test/test_httpservers.py:583]: and a test with a big file ?

File: Lib/mimetypes.py
Comment (PierreQuentel) [reply Lib/mimetypes.py]: It's not totally out of context : files with extension .json are obvious candidates for compression, this is why I have included this minor change in mimetypes, thinking that it doesn't deserve a specific PR. But if you think it's better to open another one, I will.

File: Doc/whatsnew/3.7.rst
Comment (PierreQuentel) [reply Doc/whatsnew/3.7.rst]: Same comment as below about extension .json

File: Lib/http/server.py
Comment (PierreQuentel) [reply Lib/http/server.py]: Done in  	f9963f0

File: Lib/test/test_httpservers.py
Comment (PierreQuentel) [reply Lib/test/test_httpservers.py:583]: Done in 586b8c3

File: Lib/mimetypes.py
Comment (bitdancer) [reply Lib/mimetypes.py]: Yes, any change to mimetypes should be a separate issue/PR.

File: Lib/http/server.py
Comment (vadmium) [inline Lib/http/server.py]: New version doesn’t seem to call _strip_

File: Lib/http/server.py
Comment (vadmium) [inline Lib/http/server.py]: Space is allowed between semicolon and _q_, and after the value. Also, the zero value may have a fractional decimal part.

It would be nice to use common code to parse header fields. See <https://bugs.python.org/issue3609> and <https://bugs.python.org/issue23498> about pulling a generic parser out somewhere. Perhaps the following would work.

``` Python
accept = self.headers.get_all("Accept-Encoding", ())
for accept in http.cookiejar.split_header_words(accept):
    params = iter(accept)
    encoding = next(params, "")
    for [q, v] in params:
        ...
```

Or maybe look at the _xmlrpc.server_ implementation, though that uses an ugly regular expression.

File: Lib/test/test_httpservers.py
Comment (PierreQuentel) [reply Lib/test/test_httpservers.py:583]: @matrixise 
Are you ok with the changes made since the first version, or do you see other things to do before the PR can be merged ?
If you think a separate PR is better for mimetypes I will submit it, it's just that I prefered to avoid the overhead of reviewing 2 PRs instead of one.

**Review by bitdancer — DISMISSED** [review]
I already told you that a separate *issue* and PR are required for any change to mimetypes.  Make me nosy on the issue, please (r.david.murray on the bug tracker).

**Comment by PierreQuentel** [comment]
@bitdancer 
Ok, I have removed the mimetype changes from this PR and opened an [issue](http://bugs.python.org/issue30824) on the tracker

**Comment by PierreQuentel** [comment]
@bitdancer @matrixise 
With mimetypes now out of the scope, are there other things to do before the PR can be merged ?

**Comment by PierreQuentel** [comment]
@bitdancer 
Are you ok with the changes made a few days ago ? If so, once the "Changes requested" is validated, can the PR be merged ?

**Comment by bitdancer** [comment]
My impression from scanning the issue is that it is not clear we want to accept this.

File: Misc/NEWS
Comment (terryjreedy) [inline Misc/NEWS]: NEWS has been replaced.  Add separate file with blurb.  See devguide.

File: Lib/http/server.py
Comment (vadmium) [inline Lib/http/server.py]: FYI I think it is possible to build Python without these modules

File: Lib/http/server.py
Comment (vadmium) [inline Lib/http/server.py]: See _http.client_ for an existing chunk encoder. It would be good to share code.

Anyway, I think your version will have an incorrect _0x_ prefix.

File: Lib/http/server.py
Comment (vadmium) [inline Lib/http/server.py]: _x-gzip_ is the proper name in HTTP 1.0 (which is what the class uses by default), and it is also supposed to be an alias in HTTP 1.1.

File: Lib/http/server.py
Comment (vadmium) [inline Lib/http/server.py]: Throwing away the original open file without closing it.

File: Lib/http/server.py
Comment (vadmium) [inline Lib/http/server.py]: The field value used to be a text string

Comment (PierreQuentel) [reply Lib/http/server.py]: Fixed in a7f1f47

File: Lib/http/server.py
Comment (PierreQuentel) [reply Lib/http/server.py]: Fixed in a7f1f47

File: Lib/http/server.py
Comment (PierreQuentel) [reply Lib/http/server.py]: Fixed in a7f1f47

File: Lib/http/server.py
Comment (PierreQuentel) [reply Lib/http/server.py]: I agree it would be better to share code with the chunk encoder in _http.client_, but I couldn't find a simple way to do it, and producing chunks is a very short piece of code.

I replaced `hex()` by `"{:X}".format()` ; I didn't see the bug because _http.client_ ignores the leading `0x`

File: Lib/http/server.py
Comment (PierreQuentel) [reply Lib/http/server.py]: In a7f1f47 I followed your suggestion to use `split_header_words()`. I'm not sure _http.cookiejar_ is the most natural place to put a headers parser, but it's another topic, maybe for another discussion on the issue tracker.

File: Lib/http/server.py
Comment (PierreQuentel) [reply Lib/http/server.py]: In a7f1f47 I put these imports in a `try` block and set a flag `has_gzip` ; if Python is built without these modules, compression is ignored.

File: Lib/http/server.py
Comment (PierreQuentel) [reply Lib/http/server.py]: Cf. code to parse the _Accept-Encoding_ header below.

File: Misc/NEWS
Comment (PierreQuentel) [reply Misc/NEWS]: Done in 2ca373f

File: Lib/http/server.py
Comment (vadmium) [inline Lib/http/server.py]: Seems to be unused

**Comment by PierreQuentel** [comment]
Are there others things to do before the PR can be merged ?

File: Doc/library/http.server.rst
Comment (vadmium) [inline Doc/library/http.server.rst]: I think “as _an_ argument” would be better grammar.

File: Lib/http/server.py
Comment (vadmium) [inline Lib/http/server.py]: May as well drop this import; its only function is indicating that _zlib_ is available.

File: Lib/http/server.py
Comment (vadmium) [inline Lib/http/server.py]: The documentation suggests _compressions_ can be extended. It would be good for that to work even when the _zlib_ module cannot be imported.

File: Lib/http/server.py
Comment (vadmium) [inline Lib/http/server.py]: This seems part of the _compressions_ API. Maybe better to put it in the main documentation, rather than point from the documentation to here?

File: Lib/http/server.py
Comment (vadmium) [inline Lib/http/server.py]: sorted(. . .)[-1] → max(. . .)?

File: Lib/http/server.py
Comment (vadmium) [inline Lib/http/server.py]: Why do you need the _max_ logic? Do you have a test case?

File: Lib/test/test_httpservers.py
Comment (vadmium) [inline Lib/test/test_httpservers.py]: _except Exception_, or put the specific exception in

File: Lib/test/test_httpservers.py
Comment (vadmium) [inline Lib/test/test_httpservers.py]: assertNotIn?

File: Lib/test/test_httpservers.py
Comment (vadmium) [inline Lib/test/test_httpservers.py]: Doubt _default_request_version_ is necessary

**Review by vadmium — COMMENTED** [review]
FYI once there is agreement on the behaviour, the documentation would need updating (e.g. _compressions_ dictionary vs just _gzip_ encoding, pre-computed Content-Length vs on-the-fly encoding), and the tests should be adjusted to handle missing _zlib_ and _bz2_ modules.

File: Lib/http/server.py
Comment (PierreQuentel) [reply Lib/http/server.py]: Done in baf88e4

File: Doc/library/http.server.rst
Comment (PierreQuentel) [reply Doc/library/http.server.rst]: Done in 2bce4ef

File: Lib/http/server.py
Comment (PierreQuentel) [reply Lib/http/server.py]: Done in 1bdbb3a

File: Lib/http/server.py
Comment (PierreQuentel) [reply Lib/http/server.py]: Done in 27d02bd

File: Lib/http/server.py
Comment (PierreQuentel) [reply Lib/http/server.py]: It's probably very uncommon, but an Accept-Encoding header might have several quality values for the same encoding (eg `gzip; q=0.1, gzip; q=0.5)`), or there might be several Accept-Encoding headers.

File: Lib/http/server.py
Comment (PierreQuentel) [reply Lib/http/server.py]: Done in ca68881

File: Lib/test/test_httpservers.py
Comment (PierreQuentel) [reply Lib/test/test_httpservers.py]: Done in d6157d4

File: Lib/test/test_httpservers.py
Comment (PierreQuentel) [reply Lib/test/test_httpservers.py]: Done in a456fe6

File: Lib/test/test_httpservers.py
Comment (PierreQuentel) [reply Lib/test/test_httpservers.py]: Done in 50f0a85

**Comment by PierreQuentel** [comment]
@vadmium 
I think I have made all the requested changes on the module, the test script and the documentation.

**Comment by PierreQuentel** [comment]
@rhettinger 
Do you think you will have time to review this PR on time for inclusion in Python 3.7 ?

**Comment by jerrymarino** [comment]
+1 I'd be great to see this one land 👍

**Comment by brettcannon** [comment]
Thanks for the PR, @PierreQuentel , but I'm closing this as me and various other core devs can't quite muster enough excitement about what this feature brings to warrant wanting to support it for a couple of decades.

**Comment by joshenders** [comment]
Brett, this is a capability of HTTP that has existed for decades that
Python still doesn’t support. It’s shipped with every major browser and web
server in the last 20 years. Part of the reason people are moving to Golang
is for its superior HTTP libraries. I think this PR deserves a second look.

**Comment by sijojlouis** [comment]
+1

**Comment by sijojlouis** [comment]
@PierreQuentel Great! thanks for this, I’ll try it out soon.

I have a suggestion for running (usage command) the module, I loved the simplicity of http.server - so was thinking if this could be so too, like http.eserver (e for encoded from html header or maybe c for compression).

**Pull request closed without merging**
